### PR TITLE
Future ingestion

### PR DESCRIPTION
**Pull request by bbrewer-forge**

### CODE REVIEW

File: apps/andi/lib/andi/input_schemas/form_tools.ex [hunk 1]
```diff
@@ -108,7 +108,7 @@ defmodule Andi.InputSchemas.FormTools do
     end)
   end
 
-  defp replace(map, key, function) do
+  def replace(map, key, function) do
```
Comment (bbrewer-forge) [inline apps/andi/lib/andi/input_schemas/form_tools.ex:111]: This needs to go somewhere else, as we use it everywhere

File: apps/andi/lib/andi/input_schemas/cron_tools.ex [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+  def date_and_time_to_cronstring!(date, time) do
+    time = String.pad_trailing(time, 8, ":00")
+    Timex.parse!(date <> "T" <> time, @more_forgiving_iso_basic_format)
+    |> Map.from_struct()
+    |> Map.merge(%{week: "*"})
+    |> cronlist_to_cronstring!()
+  end
+
+  defp current_year() do
```
Comment (bbrewer-forge) [inline apps/andi/lib/andi/input_schemas/cron_tools.ex:106]: cruft

File: apps/andi/lib/andi_web/input_schemas/finalize_form.ex [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        if Date.diff(dt, NaiveDateTime.utc_now()) <= 0 do
+          add_error(changeset, :date, "can't be in past")
+          |> add_error(:time, "can't be in past")
+        else
+          changeset
+        end
+    end
+  end
+
+  def replace(map, key, function) do
```
Comment (bbrewer-forge) [inline apps/andi/lib/andi_web/input_schemas/finalize_form.ex:38]: cruft

File: apps/andi/lib/andi_web/input_schemas/finalize_form.ex [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    field(:hour, :string)
+    field(:minute, :string)
+    field(:second, :string)
+  end
+
+  def changeset(%__MODULE__{} = current, changes) do
+    current
+    |> cast(changes, @cast_fields)
+    |> validate_required(@cast_fields)
+    |> validate_formats(@cast_fields, ~r|^[/.*0-9]+$|)
```
Comment (bbrewer-forge) [inline apps/andi/lib/andi_web/input_schemas/finalize_form.ex:67]: Need to test this better, though ultimately, I'll probably just use the high level validator

File: apps/andi/lib/andi_web/live/edit_live_view/finalize_form.ex [hunk 5]
```diff
@@ -162,64 +122,122 @@ defmodule AndiWeb.EditLiveView.FinalizeForm do
...
+        </form>
       </div>
     </div>
     """
   end
 
-  def handle_event("set_schedule", %{"input-field" => input_field, "value" => value}, socket) do
-    new_crontab_list = Map.put(socket.assigns.crontab_list, String.to_existing_atom(input_field), value)
-    new_cron = cronlist_to_cronstring(new_crontab_list)
+  defp timeout do
```
Comment (bbrewer-forge) [inline apps/andi/lib/andi_web/live/edit_live_view/finalize_form.ex:131]: cruft

File: apps/andi/test/unit/andi_web/input_schemas/finalize_form_test.exs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        [finalize_form(%{future_schedule: %{time: "00:01"}}), %{}],
+        [finalize_form(%{repeating_schedule: %{day: "b*"}}), %{repeating_schedule: %{day: ["has invalid format"]}}],
+        [finalize_form(%{repeating_schedule: cronlist(%{second: nil})}), %{repeating_schedule: %{second: ["can't be blank"]}}],
+        [finalize_form(), %{}],
+        [%{}, %{future_schedule: %{}, repeating_schedule: %{}}],
+      ])
+    end
+  end
+
+  defp finalize_form(overrides \\ %{}) do
```
Comment (bbrewer-forge) [inline apps/andi/test/unit/andi_web/input_schemas/finalize_form_test.exs:73]: This needs to be dried up across tests

File: apps/andi/lib/andi_web/input_schemas/finalize_form.ex [hunk 7]
```diff
@@ -0,0 +1,114 @@
+defmodule AndiWeb.InputSchemas.FinalizeForm.FutureSchedule do
```
Comment (bbrewer-forge) [inline apps/andi/lib/andi_web/input_schemas/finalize_form.ex:1]: This module needs renamed to show that it's the schema, not the form itself :|

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 8]
```diff
@@ -159,6 +162,19 @@ defmodule AndiWeb.EditLiveView do
...
   end
 
+  def handle_event("validate", %{"form_data" => form_data, "finalize_form_data" => finalize_form_data}, socket) do
+    socket = assign(socket, :finalize_form_data, finalize_form_data)
+
+    IO.inspect(finalize_form_data, label: "UP HERE?")
+
+    finalize_form_data
+    |> FinalizeForm.update_form_with_schedule(form_data)
+    |> IO.inspect(label: "do we ever get here?")
```
Comment (bbrewer-forge) [inline apps/andi/lib/andi_web/live/edit_live_view.ex:172]: ewps

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 9]
```diff
@@ -204,7 +220,9 @@ defmodule AndiWeb.EditLiveView do
     end
   end
 
-  def handle_event("save", %{"form_data" => form_data}, socket) do
+  def handle_event("save", %{"form_data" => form_data, "finalize_form_data" => finalize_form_data}, socket) do
+    socket = assign(socket, :finalize_form_data, finalize_form_data)
+
     changeset = form_data |> InputConverter.form_data_to_changeset_draft()
```
Comment (bbrewer-forge) [inline apps/andi/lib/andi_web/live/edit_live_view.ex:226]: This also needs to have the schedule form stuff plugged into it

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 10]
```diff
@@ -159,6 +162,15 @@ defmodule AndiWeb.EditLiveView do
     |> complete_validation(socket)
   end
 
+  def handle_event("validate", %{"form_data" => form_data, "finalize_form_data" => finalize_form_data}, socket) do
+    socket = assign(socket, :finalize_form_data, finalize_form_data)
```
Comment (bbrewer-forge) [inline apps/andi/lib/andi_web/live/edit_live_view.ex:166]: All of this is largely b/c you can't nest form tags, and thus, can't let the finalize form handle itself

File: apps/andi/lib/andi/input_schemas/cron_tools.ex [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    cronstring
+    |> String.split(" ")
+    |> Enum.count()
+  end
+
+  def determine_cadence_type(nil), do: determine_cadence_type("")
+  def determine_cadence_type(cadence) when cadence in ["once", "never"], do: cadence
+  def determine_cadence_type(cadence) do
+    with {:ok, parsed_cadence} <- Crontab.CronExpression.Parser.parse(cadence, true),
+         [_once] <- Crontab.Scheduler.get_next_run_dates(parsed_cadence) |> Enum.take(2) do
```
Comment (cameronmarsh) [inline apps/andi/lib/andi/input_schemas/cron_tools.ex:63]: is this enough to check that it's a future ingestion? I might be missing something but can't repeating ingestions still pass this with block

File: apps/andi/lib/andi/input_schemas/cron_tools.ex [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+  def to_repeating(type, cadence) when type in ["repeating", "future"] and cadence not in ["once", "never"], do: cronstring_to_cronlist!(cadence)
+  def to_repeating(_type, cadence), do: cronstring_to_cronlist!("0 * * * * *")
+
+  def cronlist_to_future_schedule(%{year: year, month: month, day: day, hour: hour, minute: minute, second: second} = _schedule) do
+    date = case Timex.parse("#{year}-#{month}-#{day}", @more_forgiving_iso_date_format) do
+      {:error, _} -> nil
+      {:ok, nd} -> NaiveDateTime.to_date(nd)
+    end
+
+    time = case Timex.parse("#{gee_two_zeroes(hour)}:#{gee_two_zeroes(minute)}:#{gee_two_zeroes(second)}", @more_forgiving_iso_time_format) do
```
Comment (cameronmarsh) [inline apps/andi/lib/andi/input_schemas/cron_tools.ex:81]: nice

File: apps/andi/lib/andi/input_schemas/cron_tools.ex [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+    %{date: date, time: time}
+  end
+  def cronlist_to_future_schedule(%{year: _year, month: _month, day: _day, hour: _hour, minute: _minute} = schedule) do
+    Map.put_new(schedule, :second, "0")
+    |> cronlist_to_future_schedule()
+  end
+  def cronlist_to_future_schedule(_), do: %{date: nil, time: nil}
+
+  defp gee_two_zeroes(number) do
```
Comment (cameronmarsh) [inline apps/andi/lib/andi/input_schemas/cron_tools.ex:94]: 🔥

File: apps/andi/lib/andi_web/live/edit_live_view.ex (hunk 10 above)
Comment (cameronmarsh) [reply apps/andi/lib/andi_web/live/edit_live_view.ex:166]: I will move this over if you merge before me

File: apps/andi/assets/css/finalize_form.scss [hunk 14]
```diff
@@ -21,15 +22,37 @@
 
     .finalize-form__quick-schedule {
         margin-bottom: 2rem;
+
+        input {
+            /* visibility: hidden; */
```
Comment (bbrewer-forge) [inline apps/andi/assets/css/finalize_form.scss:27]: oops

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 15]
```diff
@@ -220,12 +238,13 @@ defmodule AndiWeb.EditLiveView do
         false -> "Saved successfully. You may need to fix errors before publishing."
       end
 
-    changeset =
```
Comment (bbrewer-forge) [inline apps/andi/lib/andi_web/live/edit_live_view.ex:223]: I moved this to be part of the complete_validation stuff on line 215. NO IDEA why it was like this, but it was stomping on form changes

**Comment by bbrewer-forge** [comment]
Closing until further notice

**Pull request closed without merging**